backend/src/main/java/ai/verta/modeldb/batchProcess/CollaboratorResourceMigration.java

```diff
@@ -0,0 +1,169 @@
+    } finally {
+      ModelDBUtils.unregisteredBackgroundUtilsCount();
+    }
+
+    LOGGER.info("Migration End");
+  }
+
+  private static void migrateProjects() {
+    LOGGER.debug("Projects migration started");
+    Long count = getEntityCount(ProjectEntity.class);
+
+    int lowerBound = 0;
+    final int pagesize = CollaboratorResourceMigration.paginationSize;
+    LOGGER.debug("Total Projects {}", count);
+
+    while (lowerBound < count) {
+
+      try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
+        CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+
+        CriteriaQuery<ProjectEntity> criteriaQuery =
+            criteriaBuilder.createQuery(ProjectEntity.class);
+        Root<ProjectEntity> root = criteriaQuery.from(ProjectEntity.class);
+
+        CriteriaQuery<ProjectEntity> selectQuery =
+            criteriaQuery.select(root).orderBy(criteriaBuilder.asc(root.get("id")));
+
+        TypedQuery<ProjectEntity> typedQuery = session.createQuery(selectQuery);
+
+        typedQuery.setFirstResult(lowerBound);
+        typedQuery.setMaxResults(pagesize);
+        List<ProjectEntity> projectEntities = typedQuery.getResultList();
+
+        if (projectEntities.size() > 0) {
+          Set<String> userIds = new HashSet<>();
+          for (ProjectEntity projectEntity : projectEntities) {
+            userIds.add(projectEntity.getOwner());
+          }
+          LOGGER.debug("Project userId list : " + userIds);
+
+          // Fetch the experiment owners userInfo
+          Map<String, UserInfo> userInfoMap =
+              authService.getUserInfoFromAuthServer(userIds, null, null);
+          for (ProjectEntity project : projectEntities) {
+            WorkspaceDTO workspaceDTO =
+                roleService.getWorkspaceDTOByWorkspaceId(
+                    userInfoMap.get(project.getOwner()),
+                    project.getWorkspace(),
+                    project.getWorkspace_type());
+            // if projectVisibility is not equals to ResourceVisibility.ORG_SCOPED_PUBLIC then
+            // ignore the CollaboratorType
+            roleService.createWorkspacePermissions(
+                workspaceDTO.getWorkspaceName(),
+                project.getId(),
+                Optional.of(Long.parseLong(project.getOwner())),
+                ModelDBServiceResourceTypes.PROJECT,
+                CollaboratorTypeEnum.CollaboratorType.READ_ONLY,
+                getResourceVisibility(
+                    Optional.ofNullable(
+                        WorkspaceTypeEnum.WorkspaceType.forNumber(project.getWorkspace_type())),
+                    VisibilityEnum.Visibility.forNumber(project.getProject_visibility())));
+          }
+        } else {
+          LOGGER.debug("Total projects count 0");
+        }
+        lowerBound += pagesize;
+      } catch (Exception ex) {
+        if (ModelDBUtils.needToRetry(ex)) {
+          migrateProjects();
+        } else {
+          throw ex;
```

Comment:
I'm assuming this will make the system crash, right? If yes, approved.

Reply:
Yes, we are crashing the system